### ISSUE TITLE
Node discovery v5

### ISSUE THREAD

**Issue by holgerd77** [ethereumjs/ethereumjs-devp2p#19]
### Introduction

Node discovery v5 is on it's way, which should make it a lot easier to discover suitable nodes. Though not standardized yet, ~it would be good to have some early experimentation/draft implementation to get closer to a feeling how the protocol (could) behave~ it would be good to have this library updated with a working implementation to support the practical usage in terms of a light client implementation.

[Here](https://github.com/fjl/p2p-drafts) are some draft documents on this. Also valuable in this regard: [Talk](https://www.youtube.com/watch?v=YqGncHRtshM) from Felix at Devcon3.

Current ``py-evm`` implementation work: https://github.com/ethereum/py-evm/pull/209

### Current Situation

Currently the library supports [Node discovery v4](https://github.com/ethereum/devp2p/blob/master/discv4.md) implemented in ``src/dpt``, where [kbucket.js](https://github.com/ethereumjs/ethereumjs-devp2p/tree/master/src/dpt/kbucket.js) manages the list of peers in a ``Kademlia``-like DHT, [message.js](https://github.com/ethereumjs/ethereumjs-devp2p/tree/master/src/dpt/message.js) implements the different message types like ``ping``, ``pong`` or ``neighbours`` to discover new peers and [server.js](https://github.com/ethereumjs/ethereumjs-devp2p/tree/master/src/dpt/server.js) sets up a server to handle the discovery communication.

There are static tests for the message types in [./test/dpt-message.js](https://github.com/ethereumjs/ethereumjs-devp2p/tree/master/test/dpt-message.js), the integrated communication process is tested in [./test/integration/dpt-simulator.js](https://github.com/ethereumjs/ethereumjs-devp2p/tree/master/test/integration/dpt-simulator.js) as well as the corresponding ``eth`` and ``les`` files.

With the two ``peer-communication`` example scripts in the ``examples`` folder both ``ETH`` and ``LES`` (so full- and light client wire) communication can be tested in a real-world scenario by connecting to an actual network of in-production clients.

### Task Description

Node discovery v5 updated the structure of [Node records](https://github.com/fjl/p2p-drafts/blob/master/discv5-enr.md), brings changes to the communication flow by adding new [packet types](https://github.com/fjl/p2p-drafts/blob/master/discv5-packets.md) and introduces the concept of [topics](https://github.com/fjl/p2p-drafts/blob/master/discv5-topics.md) for exchanging on the capability of peers.

A first step into this task would be to work out some concept/idea where these additional/changed components fit into the existing v4 implementation and where current files/classes can be extended and where additional structures have to be set up. Since network communication implementation is very error-prone it will probably ease implementation to early on think along how to extend existing tests and add new tests for added functionality.

### Goals

Though being still marked as "experimental" Geth has a working v5 discovery implementation for some time. Sufficient requirement for this issue to be completed is a working ``peer-communication`` example where it is possible to set the discovery to ``v5`` - e.g. by CL parameter or example internal constant - and get a working ``LES`` and/or (optimally both) ``ETH`` connection (working means here: e.g. ``LES`` packets are actually passed through) in a reasonable amount of time (repeatedly < 5 min until first connection occurs).

Necessary side goals are:

- An extended code base which...
  - follows the style principles and coding best practices of the existing code base
  - is readable/maintainable for other developers
  - introduces no or minimal code repetitions
- An extended test suite which keeps code coverage [for the DHT](https://coveralls.io/github/ethereumjs/ethereumjs-devp2p) on more-or-less the current level
- Some additions to the ``README`` docs which gives some usage guides and reflect the API changes

### Required Skills

Everyone is invited to join and take on this task. Since this an advanced task requiring deeper knowledge of the Ethereum networking stack it is probably more likely that you succeed if you bring along skills/experience in the following fields:

- Experience with working on larger code bases in Node.js/modern ``ES6`` Javascript
- Experience with the Node.js toolchain / ecosystem
- Knowledge/understanding of generic p2p networking concepts
- Some understanding of the Ethereum networking mechanics
- Experience with assuring code quality/testing

### Guidance

For questions helping to understand the existing code base and the scope of this issue you will get active guidance by the creator of this issue (@holgerd77, EthereumJS team).

**Comment by Silur** [comment]
I guess I'll just claim this then

**Comment by holgerd77** [comment]
Cool. :-)

**Comment by holgerd77** [comment]
For reference: https://github.com/ethereum/go-ethereum/issues/2117#issuecomment-370559812
("p2p/discover: from address in ping is IPv6 unspecified in some cases")

**Comment by holgerd77** [comment]
Py-EVM implementation [in progress]: https://github.com/gsalgado/py-evm/commit/f31fbfcd654753789454863f4fb4e2c991e7af2c

**Comment by holgerd77** [comment]
@Silur Are you still open/do you have time to work on this one? Otherwise I would put this back in "help wanted", with the current comments people interested will think this is already being worked on.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 1.5 ETH (1068.95 USD @ $712.63/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereumjs/ethereumjs-devp2p/19/471).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $9,225.68 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by rauljordan** [comment]
Hey @holgerd77 interested in this, as we are currently exploring p2p in my team, Prysmatic Labs. Can you elaborate more on the requirements needed to get the bounty accepted? What would an acceptable PR look like? Any specific features you have in mind to close this?

**Comment by alexanderbez** [comment]
Interested in helping out with this as well.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.

 1. [@brianherman](https://gitcoin.co/profile/brianherman) 

 has committed to working on this project to be completed 4 weeks, 1 day from now.


@vs77bb, __please see the below comments / questions regarding approach for this ticket from the bounty hunter(s):__ 
- [@brianherman](https://gitcoin.co/profile/brianherman): 1.Read python implementation 1st week
2. Create Specification from Python Documentation.
3. Start writing javascript.

**Comment by abitrolly** [comment]
Is there an abstract describing current situation and the problem with discovery?

**Comment by holgerd77** [comment]
I'll update the issue tomorrow with a more specific work description and requirements for a bounty to be accepted.

**Comment by holgerd77** [comment]
Hi everyone, I updated the work specification, this should now hopefully accurately enough describe what is expected here, if you still have questions let me know. This might sound a bit tighter then before (didn't realize that I had this "experimental implementation" in the introduction of the issue description, I replaced this pointing to an actual working implementation).

Please re-read, take a breath and maybe sleep over one night and think, if this is a good issue for you. This is an advanced task, requiring some intrinsic motivation as well as some pre-knowledge to not make this too demanding.

Thanks everyone to expressing interest in this so far, pretty amazed about the feedback! 😄

**Comment by vs77bb** [comment]
Hi all, Vivek from Gitcoin here. It looks like @brianherman was first to 'Start Work' here on Gitcoin and thus has precedence to give this a go + @alexanderbez has offered his help as well. 

We'd encourage a collaborative effort here if it makes sense, as in line with open source ethos 😄perhaps @brianherman gives it a go and comes back around for feedback as needed?  

@rauljordan good to see you here; @holgerd77 has updated with acceptance criteria above!

**Comment by holgerd77** [comment]
@brianheram Hi Brian, does this still fit your expectations after the updated spec description and do you want to start on this? If you have any questions let me know!

**Comment by vs77bb** [comment]
@brianherman see above^ 🙂

**Comment by gitcoinbot** [comment]
@brianherman Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereumjs/ethereumjs-devp2p/19/471?snooze=1) | [3 days](https://gitcoin.co/issue/ethereumjs/ethereumjs-devp2p/19/471?snooze=3) | [5 days](https://gitcoin.co/issue/ethereumjs/ethereumjs-devp2p/19/471?snooze=5) | [10 days](https://gitcoin.co/issue/ethereumjs/ethereumjs-devp2p/19/471?snooze=10) | [100 days](https://gitcoin.co/issue/ethereumjs/ethereumjs-devp2p/19/471?snooze=100)

**Comment by vs77bb** [comment]
Hi @abitrolly @alexanderbez @rauljordan are any of you interested in taking on this bounty? @holgerd77 and I have discussed and increased the contribution to 2ETH (blockchain syncing now).

**Comment by alexanderbez** [comment]
I'm certainly happy to help on this. Seems there are three major topics: packets, node records, and topics. Perhaps we could come up with a plan/design and tackle these in parallel?

**Comment by rauljordan** [comment]
I agree that we should split this up, otherwise it would be a bunch of mega PRs. I am not as experienced on devp2p, but I'm willing to learn more and help on parts of this. Perhaps someone can come up with a proposal that can be split up into several smaller PRs/Issues?

*Raul E. Jordan, Harvard University | Ethereum Protocol Developer*

r auljordan.com
( http://rauljordan.com )
Twitter: @raulitojordan ( https://twitter.com/raulitojordan )

gpg --recv-keys 0x89725027FC8EC0D4

**Comment by holgerd77** [comment]
@alexanderbez @rauljordan Cool guys, that sounds great!

**Comment by jvmaia** [comment]
Hey i'm interested in help in this issue, i'm not experienced in devp2p but i want to learn more about.

**Comment by vs77bb** [comment]
Hi @jvmaia @rauljordan @alexanderbez I wanted to follow up as the bounty funder; is there anything I can do to help coordinate? We're happy to add to the bounty as it looks like it may become a collaborative effort.

**Comment by abitrolly** [comment]
The bounty is good, but I won't have the time to research current p2p protocols and their weaknesses.

**Comment by tcsiwula** [comment]
@holgerd77 @vs77bb @gitcoinbot can we get an extension on this? I could get it done by next week, pretty familiar with this code base -- somewhat.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 1 day, 21 hours from now. Please review their questions below:


1. [tcsiwula](https://gitcoin.co/profile/tcsiwula) has started work. 	
 * Q: If an deadline extension is made available, then I could augment the current v3 protocol to the v4.

**Comment by holgerd77** [comment]
Hi @tcsiwula, so great, happy that you want to help out, let me know if you need any assistance!

As @vs77b already stated above bounty can be paid out even after expiration:

"As an FYI, we understand the expiry date is near on Gitcoin, but will be able to pay this out upon completion, even if it takes beyond the expiration."

**Comment by vs77bb** [comment]
@holgerd77 @tcsiwula Seconding Holger, this can be paid post expiration!

**Comment by tcsiwula** [comment]
Awesome sounds good @holgerd77 @vs77bb :D ping you an update on my progress this weekend (or if I have any questions between then)!

**Comment by holgerd77** [comment]
Hi Tim @tcsiwula, don't want to pressure or something just would like to be a bit aware of the status. Have you been able to do some work on this over the weekend? Any questions?

**Comment by tcsiwula** [comment]
Thanks for the reminder! Let me get something back here within 12 hours, if not some very detailed questions~ @holgerd77

**Comment by tcsiwula** [comment]
I have to admit -- this is a very nice write up you gave 👍

**Comment by tcsiwula** [comment]
Hacking into the night on it @holgerd77 , had a bunch of meetings this afternoon

**Comment by tcsiwula** [comment]
I am falling asleep, this [PR](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42) is a WIP. The rest should come tomorrow, pinky promise o_0

**Comment by holgerd77** [comment]
Hi @tcsiwula, don't want to be too pressing on this, I know we set the original time frame a bit tight, could you nevertheless give a rough estimate on your implementation plans on the issue? Any current questions or blockers?

If you think I have conceptionalized the bounty too large in scope, feel also free to express concerns in this direction. I am also not 100% sure on this, but I am confident that we'll find a solution then.

Happy Node discovery! 😄 Holger

**Comment by ceresstation** [comment]
Hi @tcsiwula any updates here? Just want to make sure you don't get kicked off the issue. Thanks!

**Comment by tcsiwula** [comment]
Hey hey @holgerd77 @ceresstation sorry about falling off the map on this. i was working of it for quite some bit. i was traveling for a few weeks and then had a cold. it was kinda hard to gauge the current completion %, also making it backwards compatible with devp2p v4 was kind of challenging. While it was very fun to get this far I am thinking it might have been way more time than 1eth, at least for my skill level and lack of js. What do you guys think? Was I suppose to or not touch any of the ```rlpx``` files for this?

**Comment by ceresstation** [comment]
Hey @tcsiwula happy to add an additional tip if you're able to complete it promptly, @holgerd77 will have to answer regarding the `rlpx` files though.

**Comment by holgerd77** [comment]
Hi @tcsiwula, great to see that you are not completely lost without a trace!! 😄 😄 

We are actually falling here completely into the ``ETH`` volatility trap, I think when I conceptionalized this bounty, ``ETH`` price was at somewhat around 1.200 $ (*sigh*), that's why I moved to ``DAI`` for the second bounty we set up on VM testing.

I would suggest that we also move to a ``DAI`` listing here (if still possible), and - if @ceresstation can go with this - I would suggest an updated bounty here of 1.700 DAI, since the task is really somewhat demanding and - as I wrote earlier - I was not sure from the beginning if I set this up to broadly.

For this I would now nevertheless expect some constant progress on this (at least an update every two weeks) and also a somewhat more active communication if things are stalled at points. I understand your reasons but this 4 week period of non-communication was already at the very limits of what I would regard as acceptable, please at least pre-communicate with a short note in cases like these, since we also have to do some planning/estimation when we can build on-top features on this.

I am not sure if you also have to touch ``rlpx`` files for this, my first guess would be no, you can take e.g. the ``Py-EVM`` implementation I referenced above (or others) where things need to be changed. I am also not 100% sure about the need for backwards compatibility, would be interested in some current statistics on v4/v5 distribution, maybe it's also a possibility to already drop v4 support. We can figure stuff like this out during the next 2-3 weeks.

What do you think? Are these new conditions (once/if Scott approves) acceptable for you? Would still be great, if this can be finished, especially if you already did some significant progress!

Cheers
Holger

**Comment by jwasinger** [comment]
Hey @tcsiwula , if you want to pass the torch on this one, I'd be willing to take it up.  However, that would also mean removing bounty from this issue (I'm not eligible given my relationship with the Ethereum Foundation).

**Comment by ceresstation** [comment]
Hey @tcsiwula any other updates here? Otherwise we may need to move this over to @jwasinger.

**Comment by jwasinger** [comment]
Sounds good @tcsiwula .

**Comment by holgerd77** [comment]
Hi @tcsiwula, ok, we can follow on like this (though I have to admit I would be even more happy with a clear "I will do it or not" approach, but let's go on with this for now).

I think we also should be able to pay you some part of the bounty if it is not possible to finish all parts.

**Comment by vs77bb** [comment]
@holgerd77 Confirming on our end that this bounty is effectively 1,700 DAI at this point. @tcsiwula a rule of thumb we like to follow is to stay up-to-date on progress every few days, as that'll help everyone gauge current status as three weeks meanders by! Excited to see things progress. 🙂

**Comment by tcsiwula** [comment]
Sounds great @holgerd @vs77bb . I was able to confirm the following over the weekend:
 
> the python implementation is not working/finished
> the geth implementation is working but not formally specified so they can change it
> the specification is not backwards compatible nor will it ever be.
> the specification will have an upgrade path (under development/ not finished)

So this clarifies some things and helps define the bounds of expectations, at least in my mind. Removes some ambiguity about will it be able to communicate or not with other nodes (for testing/verification purposes). 

Also @holgerd77 I can commit to I will do it =) -- I was just trying to be thoughtful in case the time window was super sensitive! 😀😀😀

Now that we are all on the same page and some clarity on other clients and backwards compatibility, I will have at it. Shoot you all an update later this week (in a few days 😀).

**Comment by holgerd77** [comment]
@tcsiwula Cool! 😄 👍 That's also already valuable research work! Looking forward to the update!

**Comment by tcsiwula** [comment]
# Issue update for node discovery v5 
Completion status: **~46%** 
Today's date:  **Saturday August 25th 2018**
Target finish date:  **Monday September 3rd 2018**
Task break down: ~46% = [6.5 / 14] = [ ((4 ✅ + (3🚧/2)) ) / (14✅🚧✖️)]
Issue: #19 
Pull request: #42 
cc: @holgerd77 @vs77bb @ceresstation 

### Structural tasks
| Task name | Implemented | Task Description |
| ---------- | --------- | --------------------------------------------------- |
| node records | ✖️ No | add **node records** support for node discovery v5 |
| packet types | 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-ac6106e15ebadc071bd9547999ca99f8R171) | add **packet types** support for node discovery v5 |
| topics | 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-ac6106e15ebadc071bd9547999ca99f8R212) | add **topics** support for node discovery v5 |
| cli parameter	| ✅ [Yes](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R50) | add **v5 cli parameter** support for node discovery v5 |
| documentation	| ✅ [Yes](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-04c6e90faac2675aa89e2176d2eec7d8R65) | add **v5 cli parameter** support for node discovery v5 |

### Version 5 demo example tasks
| Demo name    									| Implemented 		|
| ------------ 									| ----------- 		|
| peer-communication-v5.js 			| 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-0e485afc4851da7ac9a6c01aff970800R5)  |
| peer-communication-les-v5.js 	| ✖️ No |

### New test coverage tasks
| Test name    | Implemented | Test Description                                   |
| ------------ | ----------- | -------------------------------------------------- |
| dpt-message-v5.js 			| ✖️ No  | static tests for v5 message types, node records, topics |
| dpt-simulator-v5.js | ✖️ No  | dpt integration communication tests for v5 |
| eth-simulator-v5.js | ✖️ No  | eth integration communication tests for v5 |
| les-simulator-v5.js | ✖️ No  | les integration communication tests for v5 |
| util.js | ✅ [Yes](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-a67f68d9ce269d1d38412c422177fecbR33) |

### Pass existing tests, style, coverage
| Name    		 | Passing 				|
| ------------ | ----------- 		|
| coveralls    | ✖️ [No](https://coveralls.io/builds/18662368)      		|
| travis       | ✅ [Yes](https://travis-ci.org/ethereumjs/ethereumjs-devp2p/jobs/420338356)      |

## update next on monday 😀
![Alt Text](https://media.giphy.com/media/GoRkPquMO1qgw/giphy.gif)

**Comment by tcsiwula** [comment]
No update for Monday. Sunday and Monday were Haskell networking. Will aim for getting half of the remaining sub tasks completed this week and update you all on Friday😃👋🏻🎉

**Comment by holgerd77** [comment]
Super-impressive summary, thanks Tim! 👍 😄

**Comment by tcsiwula** [comment]
Thanks, will have a new update in about 12 hours or so. Taking some inspiration from [geths discv5 structure](https://github.com/ethereum/go-ethereum/blob/master/p2p/discv5/topic.go).

**Comment by tcsiwula** [comment]
# Issue update for node discovery v5 

## **** latest updates **** 
I decided it would be easier to partition the version 5 logic and created a new directory `src/discv5` that mirrors `src/dpt`. Likewise for the examples, I made a separate directory and have a partial working communication example with the new packet types in `examples/discv5/simple.js`. Up next with be to complete `/discv5/node.js` and `/discv5/topic.js` and then finalize `examples/discv5/simple.js` and `examples/discv5/peer-communication.js` examples along with the tests at `/tsts/discv5/*`.

## details
Completion status: **~?46?%** 
Today's date:  **Monday September 3rd 2018**
Target finish date:  **Monday September 10th 2018**
Issue: #19 
Pull request: #42 
cc: @holgerd77 @vs77bb @ceresstation  

### Structural tasks
| Task name | Implemented | Task Description |
| ---------- | --------- | --------------------------------------------------- |
| node records | 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/2147753bea1cf1c37f25499cd73b144ddf037a18#diff-2a538f7e61c5532c588079d70256c909) | add **node records** support for node discovery v5 |
| packet types | ✅🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/2147753bea1cf1c37f25499cd73b144ddf037a18#diff-96e65b15cdd9d40b0328e0d77027f77bR257) | add **packet types** support for node discovery v5 |
| topics | 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/2147753bea1cf1c37f25499cd73b144ddf037a18#diff-273881d5364bb3d2bccc638c5374b86f) | add **topics** support for node discovery v5 |
| cli parameter	| ✅ [Yes](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R50) | add **v5 cli parameter** support for node discovery v5 |
| documentation	| ✅ [Yes](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-04c6e90faac2675aa89e2176d2eec7d8R65) | add **v5 cli parameter** support for node discovery v5 |

### Version 5 demo example tasks
| Demo name    									| Implemented 		|
| ------------ 									| ----------- 		|
/discv5/simple.js | 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/2147753bea1cf1c37f25499cd73b144ddf037a18#diff-137ee433734014114e44ee7755b96da5)|
| /discv5/peer-communication.js 			| 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/2147753bea1cf1c37f25499cd73b144ddf037a18#diff-763995f2a07a7dd541190311e0762645)  |
| /discv5/peer-communication-les.js 	| ✖️ No |

### New test coverage tasks
| Test name    | Implemented | Test Description                                   |
| ------------ | ----------- | -------------------------------------------------- |
| /tsts/discv5/dpt-message.js 			| ✖️ No  | static tests for v5 message types, node records, topics |
| /tsts/discv5/dpt-simulator.js | ✖️ No  | dpt integration communication tests for v5 |
| /tsts/discv5/eth-simulator.js | ✖️ No  | eth integration communication tests for v5 |
| /tsts/discv5/les-simulator.js | ✖️ No  | les integration communication tests for v5 |
| util.js | ✅ [Yes](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/files#diff-a67f68d9ce269d1d38412c422177fecbR33) |

### Pass existing tests, style, coverage
| Name    		 | Passing 				|
| ------------ | ----------- 		|
| coveralls    | ✖️ [No](https://coveralls.io/builds/18662368)      		|
| travis       | ✅ [Yes](https://travis-ci.org/ethereumjs/ethereumjs-devp2p/jobs/420338356)      |

**Comment by holgerd77** [comment]
Sounds good.

**Comment by tcsiwula** [comment]
Update coming tonight, hoping to have it finished today/tomorrow =)

**Comment by tcsiwula** [comment]
Yesterday I gave a go at the [discv5-enr](https://github.com/fjl/p2p-drafts/blob/master/discv5-enr.md) per [Felix's suggestion](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42#issuecomment-418909954). You can view it [here](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/ea264540fdcd9d3661197b38df84de278f97003c). Upon further reflection this morning it looks like I didn't notice the [discv4-enr-extension bridge](https://github.com/fjl/p2p-drafts/blob/master/discv4-enr-extension.md). So I will work on that today. Though I think I may put it on a new branch because it looks like it just adds two new packets ``` ENRRequest Packet (0x05)``` and ``` ENRResponse Packet (0x06)``` that do not interfere with the `v4` packet namespace. cc: @holgerd77 @vs77bb @ceresstation

**Comment by vs77bb** [comment]
Thanks again for the detailed updates here, @tcsiwula!

**Comment by holgerd77** [comment]
Hi @tcsiwula, just for information: I was somewhat unresponsive/unavailable in the last weeks due to family reasons, this is getting better now, if you need anything or are waiting for some answers let me know.

**Comment by ceresstation** [comment]
Hey @tcsiwula what's the latest update here? Anything @holgerd77 or our team can to help?

**Comment by holgerd77** [comment]
Yes, I would be interested in that as well! 😄 🐨 I think you have very well done enough overall work for  having deserved the bounty. We nevertheless have to find some consistent cut so that we can include the work you have done into the library.

So one first-thought idea would be, that we include the structural parts, mark all this as ``Experimental`` and leave all the remaining testing work for another bounty. Then we should have minimally one example working, so that it's at least possible for some general try-out.

Would this be a way to go? You can very-well make an alternative suggestion. I will also try to keep this up again and further think along within the coming 2 weeks.

Cheers
Holger

**Comment by jwasinger** [comment]
If it sounds good to you @holgerd77 @tcsiwula , I would be interested in picking up the remainder of the work on this issue.

**Comment by tcsiwula** [comment]
@ceresstation @holgerd77 yeah that sounds good to me, thank you. How should I mark them as ```Experimental```? I think I had one of the examples (/discv5/simple.js | 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/2147753bea1cf1c37f25499cd73b144ddf037a18#diff-137ee433734014114e44ee7755b96da5)|
| /discv5/peer-communication.js 			| 🚧 [Wip](https://github.com/ethereumjs/ethereumjs-devp2p/pull/42/commits/2147753bea1cf1c37f25499cd73b144ddf037a18#diff-763995f2a07a7dd541190311e0762645) ) working last but not sure if it was able to communicate properly with the corresponding geth v5 impl. Yeah that also sounds good @jwasinger .

**Comment by holgerd77** [comment]
@tcsiwula I am also not completely sure yet and have to take some more time with a closer look. Let's come up with some clear cut concept until mid next week (Wednesday, 5th), and then do eventual last necessary changes and some clean up to get things finalized and closed before Christmas. Does this make sense and is realistic? I will also have an active look during the week and propose something, think I will need 1-2 hours though.

@jwasinger You are very welcome to jump in! 👍 Just to make things clear: I have now proposed @tcsiwula the whole bounty for the work - think this is also deserved - and bounties are generally not applicable for people with a contract with EF. So this would be work within your general contract conditions.

**Comment by jwasinger** [comment]
Sounds good @holgerd77 .  It may be several weeks before I get a chance to start tackling this (so anyone is welcome to pick it up in the meantime).  Thanks for the contribution @tcsiwula . the PR looks very substantial and we very much appreciate your help.

**Comment by jwasinger** [comment]
So I realized that I actually don't have the bandwidth to take this on :/

**Comment by holgerd77** [comment]
@jwasinger Ok, thanks, feedback is really appreciated, better to realize early on then to always postpone or something! Thanks for the initial offer!

**Comment by vpulim** [comment]
> @tcsiwula Sorry I got over that one, will also circle in @vpulim since he is deeper in the the whole networking topic since he is working on this on the client. Maybe he has some good ideas on demarcation/integration of the current code state of the Discovery v5 features into the existing code base.

@holgerd77 The PR isn't ready for merging into master as-is. `src/discv5/node_record.js` has syntax errors that prevent `lib` from building and there are `console.log`s throughout the code that need to be removed. However, after commenting out the breaking code in `node_record.js` I was able to confirm the the discv4 code is still working (I was able to download blocks using the client). Without diving more deeply into the code, my initial instinct is that the PR shouldn't be merged into master until it's cleaned up and tested more thoroughly. Also, I would recommend holding off until the v5 spec is finalized or a real urgent need for v5 arises. Currently, discv4 seems to be working just fine for client needs.

**Comment by holgerd77** [comment]
Ok, thanks @vpulim for the analysis, then let's do the following:

@tcsiwula I've send you a collaborator invite for the repository. Can you do some last basic cleanups mentioned in the comment above and then do a final re-pushing directly to the repository so that we have administrative control over this?

I would then say that we'll leave this as an open PR as a future working basis for now and I would recommend @ceresstation that we pay out a bounty of 1.000 DAI for this (with the lesson also learned to never again do ``ETH`` bounties). How does this sound for everyone?

**Comment by holgerd77** [comment]
reminder ping (see message above): @ceresstation @tcsiwula

**Comment by ceresstation** [comment]
@tcsiwula is there anything I can do on my end to help get this across the finish line? It's been outstanding for quite a while.

**Comment by holgerd77** [comment]
Hi @tcsiwula, see the edited comment from @ceresstation above with a note on adjustment of the bounty value. Please let me know if it is possible to do the last adjustments on this I mentioned above until the end of the year, of if you have a significant reason why this is not possible (you can also PM me on Gitter). I would otherwise close/cancel this bounty.

**Comment by tcsiwula** [comment]
Hello @holgerd77 @ceresstation sorry have been behind on my email. Let me process the above thread. So was the amount 1.000 DAI equal to $1000 or what was the USD value?

**Comment by holgerd77** [comment]
1 DAI = 1 US $, so yes.

**Comment by holgerd77** [comment]
Ok, just re-read the thread, seems that we have two a bit conflicting statements here with @ceresstation proposing a 2 ETH adjustment and me some conversion to 1.000 DAI.

I would find the 1.000 DAI version a bit more appropriate TBH, since we would leave this task in a very open state with the suggestion I made above, this has exceeded the originally planned time by some order of magnitude and lastly the bounty actually had been defined in ETH with everyone knowing about the risk of ETH price fluctuation.

Anyhow, it's Christmas 🎄 😄 and at the end it's up to Gitcoin to decide on the bounty level (since they have to pay! 😄). So let's take the 2 ETH adjustment version, this would be 712,63 $/ETH (price at bounty opening, see [Gitcoin Bot comment](https://github.com/ethereumjs/ethereumjs-devp2p/issues/19#issuecomment-390484928) above) * 2 ETH = 1.425,26 $ ~ **1.425 DAI**.
(I would now really stick to DAI here, since ETH price is very fluctuant right now and we very well might end up with another 2x difference in 2 weeks or so).

I will stick to the end-of-year deadline though.

So please do the following:

- Do a new PR with your work directly to the original repository so that we gain administrative control on the work
- Do the syntax error and ``console.log()`` cleanup mentioned in the [comment](https://github.com/ethereumjs/ethereumjs-devp2p/issues/19#issuecomment-445544936) from @vpulim above
-  Optimally make sure that we can at least run one example to test the new functionality (won't make this a blocker though)

We can then pay out the bounty.

Best
Holger

**Comment by ceresstation** [comment]
Hey @holgerd77 sorry we're on the same page, I was just saying I'd be willing to pay a bounty amount equal to the amount 2 ETH was worth when the bounty began *in DAI*. In other words I can confirm that we'd be able to pay $1,425 DAI. @tcsiwula please confirm that this is good on your end :)

**Comment by tcsiwula** [comment]
Okay sounds good to me @holgerd77 @ceresstation will get it done by Sunday for review!

**Comment by tcsiwula** [comment]
working on right now, a bit late but expect soon.

**Comment by holgerd77** [comment]
That's ok.

**Comment by tcsiwula** [comment]
Just wrapping up. Created a new PR & branch #48  [summary here](https://github.com/ethereumjs/ethereumjs-devp2p/pull/48) compiles and runs now successfully should be okay to close out the issue. Let me know if you need me to submit or close the bounty manually @holgerd77 @ceresstation 

Further work still needs to be done with regards to the Node Record Specification and testing. I can try to work on this some more in 2019 or at least define some subset pr's. Kinda in the middle of moving, interviewing and spending holidays with my family (just a fyi). Normally I would be more responsive!

Cheers~ 🤓😄

**Comment by holgerd77** [comment]
We can very well see this as accomplished, Tim @tcsiwula thanks for doing all the work on this, we've learned to do more fine-grained Gitcoin issues in the future. Hope it was not too overwhelming, would be great if you can continue with some work in the time to come. 😄 👍

@ceresstation Can you initiate the bounty payout (the last comment I made on the PR shouldn't be a blocker, this is just some test run fixes and the PR will stay open anyhow)?

**Comment by ceresstation** [comment]
@tcsiwula thanks for getting that done, @holgerd77 initiating the payout now!

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 2.00000  ETH (301.9 USD @ $150.95/ETH) has been granted to @tcsiwula for this issue from @vs77bb. ⚡️ 

Nice work @tcsiwula! Your tip has automatically been deposited in the ETH address we have on file.

 * $47669.74 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 700.00000  DAI (700.0 USD @ $1.0/DAI) has been granted to @tcsiwula for this issue from @vs77bb. ⚡️ 

Nice work @tcsiwula! Your tip has automatically been deposited in the ETH address we have on file.

 * $47682.59 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by vs77bb** [comment]
@tcsiwula Please confirm you've received the payment! I remember when this task began - glad to see it across the finish line. Thanks for your patient and thoughtful reviews along the way, @holgerd77 🙂

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 425.00000  DAI (425.0 USD @ $1.0/DAI) has been granted to @tcsiwula for this issue from @ceresstation. ⚡️ 

Nice work @tcsiwula! Your tip has automatically been deposited in the ETH address we have on file.

 * $46136.83 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by kuhnchris** [comment]
Hi there @tcsiwula , would you mind finishing this on the gitcoin side aswell, so we can finish this bounty? Thank you!

Chris

**Comment by tcsiwula** [comment]
Pretty sure got the payout, will confirm when I get access to my desktop next. Also will close out on gitcoin side soon!
